### ISSUE TITLE
Issues with Cypress component testing (Angular)

### ISSUE THREAD

**Issue by rkrisztian** [jscutlery/test-utils#72]
I've run into some problems in my [test project](https://github.com/rkrisztian/cypress-jest-test-project/commits/router-test) (branch: `router-test`, as linked).

### 1. The app is routed to `/__`

With `npx nx serve`, the app gets routed to `/my-page`, as expected. But when running `app.component.cy-spec.ts`, the default route is `/__` for some reason. Do you know why that happens and how to get around it?

### 2. Assets not loaded

The `app.component.cy-spec.ts` shows that Cypress component tests cannot load any assets, like `nx-logo.png` in my example. This can get even worse when a component needs the language translation files like `en.json` or something like that (typically used with `ngx-translate`).

### 3. Cannot get access to component instance

One of the lines I added into `app.component.cy-spec.ts` is this:

```
cy.get('h1').contains(`Welcome to ${new AppComponent().title}`);
```

This is a too contrived example of course, but with Cypress, I get the impression that component testing is closer to a black-box testing approach. With `TestBed`, I can access component fields/methods too. But with Cypress, this is much more limited. E.g. how do I talk to component `@Input`s and `@Output`s? Do I need to replace the `MyComponent` in my `mount` call with `<my-component ...></my-component>` instead? Which I personally find a bit cumbersome, not so easy to communicate with the component this way.

Thank you for your help.

**Comment by rkrisztian** [comment]
For problem 1 I found a solution:

```diff
diff --git a/apps/test-project/src/app/app.component.cy-spec.ts b/apps/test-project/src/app/app.component.cy-spec.ts
index 8821229..293d044 100644
--- a/apps/test-project/src/app/app.component.cy-spec.ts
+++ b/apps/test-project/src/app/app.component.cy-spec.ts
@@ -1,10 +1,16 @@
 import { mount } from '@jscutlery/cypress-angular/mount';
 import { AppComponent } from "./app.component";
 import { AppModule } from "./app.module";
+import { APP_BASE_HREF } from "@angular/common";
 
 describe('app component', () => {
   it('should say welcome', () => {
-    mount(AppComponent, {imports: [AppModule]});
+    mount(AppComponent, {
+      imports: [AppModule],
+      providers: [
+        {provide: APP_BASE_HREF, useValue: '/__'}
+      ]
+    });
     cy.get('h1').contains(`Welcome to ${new AppComponent().title}`);
   });
 });
```

But this should be provided by the library, I think.

**Comment by rkrisztian** [comment]
For problem 2, I've been experimenting with the following:

```typescript
import * as enJson from '../../assets/i18n/en.json';

class FakeLoader implements TranslateLoader {
  getTranslation(lang: string): Observable<any> {
    return of(enJson);
  }
}

describe('MyComponent', () => {
  it('shows the title', () => {
    mount(MyComponent, {
      imports: [
        TranslateModule.forRoot({
          loader: {
            provide: TranslateLoader,
            useClass: FakeLoader,
            deps: [HttpClient]
          },
        }),
        HttpClientModule,
        AppRoutingModule
      ],
      providers: [{ provide: APP_BASE_HREF, useValue: '/__' }]
    });
    cy.get('h1').contains('My title');
  });
});
```

This code has the following problems:

* I am still lucky I could resolve assets that way. Images are still not loaded.
* If this component manages its own translation:
    * If the title is contained in a child component, how do I tell Cypress to load it?
* Otherwise:
    * How do I inject services like `TranslateService`, into my test case, so that I can perform the translation? 

Basically I wanted to do something like this:
```typescript
import { getTestBed } from '@angular/core/testing';
import { TranslateService } from '@ngx-translate/core';

const translate = getTestBed().get(TranslateService);
translate.use('en');
```

(The details are more complicated, see https://github.com/ngx-translate/core/issues/636#issuecomment-334648015, which also mentions `TranslateTestingModule`, which still requires Angular 10.)

**Comment by rkrisztian** [comment]
Regarding problem 3, I think we should be careful with access to the component instance. The Angular testing library [distinguishes two kinds of components](https://github.com/testing-library/angular-testing-library#guiding-principles): one kind used for rendering, those should be driven through the DOM. The others, are probably not good candidates for Cypress component testing. That raises the question: are mixed components a bad design, mixing responsibilities? Or can there be special cases that are acceptable? I'm still investigating this question.

**Comment by rkrisztian** [comment]
For problem 3, while I think that accessing component fields or methods may be a bad practice due to testing implementation details, and the Angular testing library agrees that [if we render a component, we should be able to test it only through the DOM](https://github.com/testing-library/angular-testing-library#guiding-principles); there may be exceptional cases: [an article](https://dev.to/julianobrasil/accessing-angular-components-methods-from-within-cypress-4imf) suggests that sometimes acess to the component instance our best option, e.g. rather than dealing with e.g. the subtleties of a canvas.

**Comment by azaeng04** [comment]
Eish now testing Angular services with Cypress is what I need.

**Comment by azaeng04** [comment]
@rkrisztian I am using https://github.com/bahmutov/cypress-angular-unit-test mount capability and jscutlery's https://github.com/jscutlery/test-utils/blob/main/packages/cypress-angular-dev-server dev server.

With the existing cypress-angular-unit-test framework there is the capability to make use of TestBed.

**Comment by rkrisztian** [comment]
@azaeng04 , thanks, but it feels like a risky combination of libraries to me in terms of future proofness. I would rather minimize my dependencies, and not only _cypress-angular-dev-server_ is internal, but also _cypress-angular-unit-test_ is supposed to be replaced by this more official package.

**Comment by azaeng04** [comment]
Yeah @rkrisztian completely agree. I was just playing around with it, but does not seem like it would be a long term sustainable solution.
